gapis/api/vulkan/frame_loop.go

```diff
@@ -126,129 +123,198 @@ func newFrameLoop(ctx context.Context, c *capture.GraphicsCapture, numInitialCmd
 		imageChanged:   make(map[VkImage]bool),
 		imageDestroyed: make(map[VkImage]bool),
 		imageToBackup:  make(map[VkImage]VkImage),
-	}
 
-	f.loopStartCmd, f.loopEndCmd = f.getLoopStartAndEndCmd(ctx, Cmds)
+		descriptorSetLayoutCreated: make(map[VkDescriptorSetLayout]bool),
+		// descriptorSetLayoutChanged:   make(map[VkDescriptorSetLayout]bool),
+		descriptorSetLayoutDestroyed: make(map[VkDescriptorSetLayout]bool),
+		// descriptorSetLayoutToBackup:  make(map[VkDescriptorSetLayout]VkDescriptorSetLayout),
+	}
 
+	f.loopStartIdx, f.loopEndIdx = f.getLoopStartAndEndIndices(ctx, Cmds)
 	return f
 }
 
-func (f *frameLoop) Transform(ctx context.Context, id api.CmdID, cmd api.Cmd, out transform.Writer) {
+func (f *frameLoop) Transform(ctx context.Context, cmdId api.CmdID, cmd api.Cmd, out transform.Writer) {
+
 	ctx = log.Enter(ctx, "FrameLoop Transform")
 
-	if cmd == f.loopStartCmd {
-		log.D(ctx, "Loop: start loop at frame %v, id %v, cmd %v.", f.frameNum, id, cmd)
+	if cmd == f.cmds[f.loopStartIdx] {
+
+		log.D(ctx, "Loop: start loop at frame %v, cmdId %v, cmd %v.", f.frameNum, cmdId, cmd)
 		f.detectChangedResource(ctx, out.State())
 
-		st := GetState(out.State())
-		sb := st.newStateBuilder(ctx, newTransformerOutput(out))
-		defer sb.ta.Dispose()
+		apiState := GetState(out.State())
 
-		if err := f.backupChangedResources(ctx, sb); err != nil {
+		stateBuilder := apiState.newStateBuilder(ctx, newTransformerOutput(out))
+		defer stateBuilder.ta.Dispose()
+
+		if err := f.backupChangedResources(ctx, stateBuilder); err != nil {
 			log.E(ctx, "Failed to backup changed resources: %v", err)
 			return
 		}
-		// Add jump label
-		sb.write(sb.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
+
+		stateBuilder.write(stateBuilder.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
 			f.loopCountPtr = b.AllocateMemory(4)
 			b.Push(value.S32(f.loopCount))
 			b.Store(f.loopCountPtr)
 			b.JumpLabel(uint32(0x1))
 			return nil
 		}))
+
 		out.NotifyPreLoop(ctx)
 
-	} else if cmd == f.loopEndCmd {
-		log.D(ctx, "Loop: last frame is %v id %v, cmd is %v.", f.frameNum, id, cmd)
-		st := GetState(out.State())
-		sb := st.newStateBuilder(ctx, newTransformerOutput(out))
-		defer sb.ta.Dispose()
-		out.MutateAndWrite(ctx, id, cmd)
+	} else if cmd == f.cmds[f.loopEndIdx] {
+
+		log.D(ctx, "Loop: last frame is %v cmdId %v, cmd is %v.", f.frameNum, cmdId, cmd)
+
```

Comment:
What's the value in having so empty blank lines in this body?

Reply:
I've tried to use white space to group things functionally across the whole file.  The code in this block mostly kicks off a series of high level operations via single functions and thus many of the groups are a single line. Though you will still see some grouping like lines 172/173, and 180-183. There is no semantic meaning to this, but I find white space makes code look a lot less intimidating and easier on the eyes. Also when a block boils down to kicking off a series of high level operations I find it draws the eyes to the fact quite dramatically.